src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsBufferAlloc.cs

```diff
@@ -2,37 +2,104 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-using System;
-using System.Diagnostics;
+using System.Buffers;
+using System.Threading.Tasks;
 
 namespace Microsoft.Data.SqlClientX.IO
 {
     /// <summary>
     /// Provides buffer management for reader/writer implementations
     /// </summary>
-    internal class TdsBufferAlloc 
+    internal class TdsBufferManager
     {
-        private readonly byte[] _buffer;
+        private readonly ArrayPool<byte> _bufferPool = ArrayPool<byte>.Shared;
+
+        private byte[] RentBuffer(int size) => _bufferPool.Rent(size);
+
+        private void ReturnBuffer(byte[] buffer) => _bufferPool.Return(buffer, clearArray: true);
```

Comment:
Are you sure you want to always clear the buffer, it's usually unnecessary overhead for byte buffers.